### PR TITLE
Enhance security

### PR DESCRIPTION
**Pull request by thecavemanlogic**
When a new user is created, `User.id` is generated by `uuid4()` if the `User` constructor is called without an `id` argument. Older users where `id` and `username` share the same value are kept as-is.

A new cookie, `id`, was added to store `User.id`. The `user` cookie now stores `User.username`. All requests now process data based on the new `id` cookie instead of the `user` cookie.

How does this look?

### CODE REVIEW

**Review by sschaub — CHANGES_REQUESTED** [review]
This line of code appears many times:
```
user = User.get(request.COOKIES['id'])
```
I would like to see this refactored a bit to abstract the business of accessing the cookie ID. I suggest creating a method User.getCurrent(request) that has
```
return User.get(request.COOKIES['id']) if request.COOKIES.get('id') else None
```
or something along those lines.

File: opencontest/contest/models/user.py [hunk 1]
```diff
@@ -28,13 +30,23 @@ def getByName(username: str):
...
+
     def save(self):
-        if self.id == None:
-            if self.username in [user.id for user in User.all()]:
-                return  # don't create duplicate user
-            self.id = self.username
+        # If a user with a username of self.username
+        # does not exists, then add it to the global
+        # data structures
+        if self.id not in [user.username for user in User.all()]:
```
Comment (sschaub) [inline opencontest/contest/models/user.py]: I am confused by this code. Why are we checking to see if self.id is not one of the user's usernames?

File: opencontest/contest/pages/correctlog.py [hunk 2]
```diff
@@ -26,7 +26,7 @@ def constructTableRows(listOfSubmissions):
 
 
 def generateLogReport(request):
-    user = User.get(request.COOKIES['user']) if request.COOKIES.get('user') else None
+    user = User.getCurrent(request) if request.COOKIES.get('id') else None
```
Comment (sschaub) [inline opencontest/contest/pages/correctlog.py]: Why not just:
```
user = User.getCurrent(request)
```

File: opencontest/contest/models/user.py [hunk 3]
```diff
@@ -8,7 +8,9 @@
 
 class User:
     def __init__(self, username: str, fullname: str, password: str, type: str, id: str = None):
-        self.id = id
+        # Use uuid4 as an id because it is hard to brute
+        # force
+        self.id = id or str(uuid4())
```
Comment (sschaub) [inline opencontest/contest/models/user.py]: It is helpful when inspecting database files to have the username as part of the id. Maybe add it as a prefix, like this?
```
self.id = id or (username + str(uuid4()))
```

**Review by sschaub — CHANGES_REQUESTED** [review]

File: opencontest/contest/models/user.py (hunk 1 above)
Comment (thecavemanlogic) [reply opencontest/contest/models/user.py]: That is confusing... That `self.id` should be `self.username`. I'll go and fix that.

**Comment by thecavemanlogic** [comment]
I rewrote the `User.save()` function so it is more clear what the method is doing and fulfilled your other requests.

**Review by sschaub — APPROVED** [review]
Looks good!

**Pull request merged**